### PR TITLE
MotoROS v1.7.0

### PR DESCRIPTION
**Pull request by ted-miller**
- Implement velocity feedback for DX200 and YRC1000 controllers.
- Remove unneeded validation check from RosSetupValidation.xLib.
- Fix starting position check for SLUBT robots.

### CODE REVIEW

**Comment by ted-miller** [comment]
Eric Marcil will do additional testing this week.

**Comment by ted-miller** [comment]
Updated based on testing by Marcil.

File: motoman_driver/MotoPlus/Controller.c [hunk 1]
```diff
@@ -227,7 +227,7 @@ BOOL Ros_Controller_Init(Controller* controller)
 		if(grpNo < controller->numGroup)
 		{
 			// Determine if specific group exists and allocate memory for it
-			controller->ctrlGroups[grpNo] = Ros_CtrlGroup_Create(grpNo, controller->interpolPeriod);
+			controller->ctrlGroups[grpNo] = Ros_CtrlGroup_Create(grpNo, (grpNo==(controller->numGroup-1)), controller->interpolPeriod);
```
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/Controller.c]: Could this line get a bit more of a comment?

It probably becomes clearer when looking up the declaration of `Ros_CtrlGroup_Create(..)`, but what is the boolean there?

File: motoman_driver/MotoPlus/CtrlGroup.c [hunk 2]
```diff
@@ -31,20 +31,6 @@
...
-CtrlGroup* Ros_CtrlGroup_Create(int groupNo, float interpolPeriod);
-BOOL Ros_CtrlGroup_GetPulsePosCmd(CtrlGroup* ctrlGroup, long pulsePos[MAX_PULSE_AXES]);
-BOOL Ros_CtrlGroup_GetFBPulsePos(CtrlGroup* ctrlGroup, long pulsePos[MAX_PULSE_AXES]);
-void Ros_CtrlGroup_ConvertToRosPos(CtrlGroup* ctrlGroup, long pulsePos[MAX_PULSE_AXES], 
-									float rosPos[MAX_PULSE_AXES]);
-void Ros_CtrlGroup_ConvertToMotoPos(CtrlGroup* ctrlGroup, float rosPos[MAX_PULSE_AXES],
-									long pulsePos[MAX_PULSE_AXES]);
-UCHAR Ros_CtrlGroup_GetAxisConfig(CtrlGroup* ctrlGroup);
-BOOL Ros_CtrlGroup_IsRobot(CtrlGroup* ctrlGroup);
-
```
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/CtrlGroup.c:47]: These have moved to the header?

File: motoman_driver/MotoPlus/CtrlGroup.c [hunk 3]
```diff
@@ -367,13 +353,22 @@ BOOL Ros_CtrlGroup_GetFBServoSpeed(CtrlGroup* ctrlGroup, long pulseSpeed[MAX_PUL
...
+		//combine both registers into single 4 byte value (0.0001 deg/sec or 1 um/sec)
 		dblRegister = (registerValuesLong[(i * 2) + 1] << 16) | registerValuesLong[i * 2];
 
 		//convert to pulse/sec
-		dblRegister /= 10000.0; //deg/sec
-		dblRegister *= RAD_PER_DEGREE; //rad/sec
-		dblRegister *= ctrlGroup->pulseToRad.PtoR[i]; //pulse/sec
+		if (ctrlGroup->axisType.type[i] == AXIS_ROTATION)
+		{
+			dblRegister /= 10000.0; //deg/sec
```
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/CtrlGroup.c]: Suggestion: use `1e4` (and `1e5` below)?

Or: `1e-4` (and `1e-5`) and multiply.

Comment (ted-miller) [reply motoman_driver/MotoPlus/CtrlGroup.c]: Ok

File: motoman_driver/MotoPlus/CtrlGroup.c (hunk 2 above)
Comment (ted-miller) [reply motoman_driver/MotoPlus/CtrlGroup.c:47]: Previously, they were located in both the header and the source.  I removed the duplication.

File: motoman_driver/MotoPlus/Controller.c (hunk 1 above)
Comment (ted-miller) [reply motoman_driver/MotoPlus/Controller.c]: New commit has additional comments.

**Comment by gavanderhoorn** [comment]
Just tested this on our FS100. No regressions as far as I've tested.

I cannot test velocity reporting, nor the new SLUBT starting position check.

**Comment by ted-miller** [comment]
The latest SLUBT update has been tested on a live controller in simulation mode. It has not been tested on physical arm.  (But I am very confident that it will be ok.)

**Comment by gavanderhoorn** [comment]
@ted-miller wrote:
> tested on a live controller in simulation mode.

you'll have to teach me how to do that one day.

**Review by gavanderhoorn — APPROVED** [review]
Changes have been tested as best we can, let's iterate on improvements in new PRs.

**Comment by gavanderhoorn** [comment]
I'll wait on Travis to complete and then merge.

💯 @ted-miller: this is a really nice implementation of velocity feedback I believe.

**Comment by gavanderhoorn** [comment]
This would be something for a follow-up PR, but: would there be some way to implement a UI / settings panel of some sorts using MotoPlus? You mentioned in a private email that the implementation of the velocity feedback is quite 'resource intensive' (M-registers) and that it could complicate integration of MotoROS with other jobs on the controller.

Would it be an idea to give users a choice, so that they can disable the feedback if they're not interested in it and avoid the resource usage (which would be unneeded in that case)?

**Comment by ted-miller** [comment]
> > TM: tested on a live controller in simulation mode.
>
> GH: you'll have to teach me how to do that one day.

Unfortunately, the Yaskawa password is required to reconfigure your controller for a different arm.

**Comment by ted-miller** [comment]
> would there be some way to implement a UI / settings panel of some sorts

This is definitely possible. There is an SDK available to develop a custom GUI application on the programming pendant. The development kit is a paid option, but the compiled binary and installer could be posted online.

Let me know if you have some ideas for things that would be good to have configurable, beyond just the velocity feedback. Maybe enable/disable torque feedback? Or display a diagnostic monitor screen? Perhaps display all the controller information that is calculated at startup (max speeds, pulse->radian, etc)?

**Comment by gavanderhoorn** [comment]
@ted-miller: I've opened #193 to discuss the UI.

**Comment by gavanderhoorn** [comment]
> > > TM: tested on a live controller in simulation mode.
> >
> > GH: you'll have to teach me how to do that one day.
> 
> Unfortunately, the Yaskawa password is required to reconfigure your controller for a different arm.

so ... ? ;)

**Comment by gavanderhoorn** [comment]
This is a very nice addition to MotoROS @ted-miller. Thanks for this.

Now all I need to do is find a DX200/YRC1000 somewhere ..

**Pull request merged**